### ISSUE TITLE
[koffi] Usage with Jest

### ISSUE THREAD

**Issue by CommandMC** [Koromix/rygel#51]
Hello there! I'm trying to use Koffi together with [Jest](https://jestjs.io), but it seems like Jest is keeping some sort module state alive between tests, causing aliased types to stay defined after a test is done running.
I'm suspecting that this is caused by Jest re-using one Node process for multiple tests

For more details on how exactly I'm using Koffi (and Jest), see https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/2717
I haven't setup a minimal reproducible example yet, if that would make it easier, let me know

**Comment by Koromix** [comment]
Indeed, the custom types are shared across the Node instance. Which means that if you try to create a type a second time (e.g. a struct), it fails with a "Duplicate name" exception. same thing for type aliases.

Koffi works this way to allow the prototype parser to work. When you write `lib.func('stuff *new_stuff();');`, Koffi has to find the `stuff` type somewhere and that's why if was done this way. Also, unlike functions, you may want to share use the same type with different libraries, that's why it's done in a shared global namespace.

I could add a `clear` function to reset all internal states, mainly for testing purposes, but I don't like that. Maybe I can introduce support for separate namespaces (if needed), and keep the default namespace as-is. I'll try something like this soon.

**Comment by Koromix** [comment]
Mmmh, now that I think about it, I think the simplest solution is to add a setting to allow types to be redefined (not allowed by default but you could change it).

That's what I'll start with, I think.

**Comment by CommandMC** [comment]
Thanks! A dedicated `clear()` function to use specifically in this scenario doesn't sound too bad IMO, but of course that decision is up to you

If there's anything to test (or more info is needed), let me know

**Comment by CommandMC** [comment]
On another note, I also ran into some issues using this library in a Flatpak container & opened up https://github.com/flatpak/flatpak/issues/5420. Just in case you have any input on why *that* issue might be happening
